File: articles/calculus-for-systems-modeling/articles/chaos-and-sensitivity-to-initial-conditions/c/chaos_sensitivity_audit.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>

double logistic_map(double x, double r){ return r*x*(1.0-x); }

int main(void){
  double r = 3.9;
  double x_reference = 0.2;
  double x_perturbed = 0.2 + 1e-8;
  printf("step,x_reference,x_perturbed,absolute_difference,log_difference,warning\n");
  for(int step=0; step<=100; step++){
    double difference = fabs(x_reference - x_perturbed);
    double log_difference = difference > 0.0 ? log(difference) : 0.0;
    printf("%d,%.12f,%.12f,%.12g,%.12f,Trajectory divergence depends on parameter value initial uncertainty numerical precision and iteration count.\n", step, x_reference, x_perturbed, difference, log_difference);
    x_reference = logistic_map(x_reference, r);
    x_perturbed = logistic_map(x_perturbed, r);
  }
  return EXIT_SUCCESS;
}

File: articles/calculus-for-systems-modeling/articles/chaos-and-sensitivity-to-initial-conditions/fortran/chaos_sensitivity_audit.f90
program chaos_sensitivity_audit
  implicit none
  integer :: step
  real(8) :: r, x_reference, x_perturbed, difference, log_difference
  r = 3.9d0
  x_reference = 0.2d0
  x_perturbed = 0.2d0 + 1.0d-8
  print '(A)', 'step x_reference x_perturbed absolute_difference log_difference'
  do step = 0, 100
    difference = abs(x_reference - x_perturbed)
    if (difference > 0.0d0) then
      log_difference = log(difference)
    else
      log_difference = 0.0d0
    end if
    print '(I6,4F18.10)', step, x_reference, x_perturbed, difference, log_difference
    x_reference = logistic_map(x_reference, r)
    x_perturbed = logistic_map(x_perturbed, r)
  end do
contains
  real(8) function logistic_map(x, r)
    real(8), intent(in) :: x, r
    logistic_map = r * x * (1.0d0 - x)
  end function
end program
